akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashManagement.scala

```diff
@@ -45,34 +62,75 @@ private[akka] trait StashManagement[C, E, S] {
+        "Unstashing message from {} stash: [{}]",
+        if (buffer eq stashState.internalStashBuffer) "internal" else "user", buffer.head)
+
+      if (stashState.isUnstashAllInProgress)
+        stashState.decrementUnstashAllProgress()
+
+      buffer.unstash(setup.context, behavior, 1, ConstantFun.scalaIdentityFunction)
     } else behavior
+
+  }
+
+  /**
+   * Subsequent `tryUnstashOne` will first drain the user stash buffer, before using the
+   * internal stash buffer. It will unstash as many commands as are in the buffer when
+   * `unstashAll` was called, i.e. if subsequent commands stash more those will
+   * not be unstashed until `unstashAll` is called again.
+   */
+  protected def unstashAll(): Unit = {
+    if (stashState.userStashBuffer.nonEmpty) {
+      if (setup.settings.logOnStashing) setup.log.debug(
+        "Unstashing all [{}] messages from user stash, first is: [{}]",
+        stashState.userStashBuffer.size, stashState.userStashBuffer.head)
+      stashState.startUnstashAll()
+      // tryUnstashOne is called from EventSourcedRunning at the end of processing each command (or when persist is completed)
+    }
   }
 
+  protected def isUnstashAllInProgress: Boolean =
+    stashState.isUnstashAllInProgress
+
 }
 
-/**
- * INTERNAL API
- * Main reason for introduction of this trait is stash buffer reference management
- * in order to survive restart of internal behavior
- */
-@InternalApi private[akka] trait StashReferenceManagement {
+/** INTERNAL API: stash buffer state in order to survive restart of internal behavior */
+@InternalApi
+private[akka] class StashState(settings: EventSourcedSettings) {
 
-  private var stashBuffer: OptionVal[StashBuffer[InternalProtocol]] = OptionVal.None
+  private var _internalStashBuffer: StashBuffer[InternalProtocol] = StashBuffer(settings.stashCapacity)
+  private var _userStashBuffer: StashBuffer[InternalProtocol] = StashBuffer(settings.stashCapacity)
+  private var unstashAllInProgress = 0
 
-  def stashBuffer(settings: EventSourcedSettings): StashBuffer[InternalProtocol] = {
-    val buffer: StashBuffer[InternalProtocol] = stashBuffer match {
-      case OptionVal.Some(value) ⇒ value
-      case _                     ⇒ StashBuffer(settings.stashCapacity)
-    }
-    this.stashBuffer = OptionVal.Some(buffer)
-    stashBuffer.get
+  def internalStashBuffer: StashBuffer[InternalProtocol] = _internalStashBuffer
+
+  def userStashBuffer: StashBuffer[InternalProtocol] = _userStashBuffer
+
+  def clearStashBuffers(): Unit = {
+    _internalStashBuffer = StashBuffer(settings.stashCapacity)
+    _userStashBuffer = StashBuffer(settings.stashCapacity)
+    unstashAllInProgress = 0
   }
 
-  def clearStashBuffer(): Unit = stashBuffer = OptionVal.None
+  def isUnstashAllInProgress: Boolean =
+    unstashAllInProgress > 0
+
+  def decrementUnstashAllProgress(): Unit =
+    unstashAllInProgress -= 1
+
+  def startUnstashAll(): Unit =
+    unstashAllInProgress = _userStashBuffer.size
+
 }
```

Comment:
This class has so many nested if functions it indicates the design of it internally should be cleaner. Maybe leveraging anonymous functions.